### DISCUSSION TITLE
How to do "theming" with ACSS?

### DISCUSSION THREAD

**Discussion by thierryk** [acss-io/atomizer#434] [Q&A] [answered]
# How to do "theming" with ACSS?

## Question from [@jitendravyas](https://github.com/jitendravyas) via Gitter

Atomizer config is not able to do this:

```js
{
  "custom": {
    "Color-red": "#FF0000",
    "Danger": "Color-red"
  }
}
```

I think real time themeable website is not possible with Atomizer.

**Comment by thierryk** [comment, marked as answer]
For theming, it is best to leverage *custom properties* (aka CSS variables) and that's **really easy to do with Atomizer**.

In the config:

```js
{
  "custom": {
    "primaryColor": "var(--primaryColor)",
    "secondaryColor": "var(--secondaryColor)"
  }
}
```

In the markup:

```html
<div class="C(primaryColor) Bgc(secondaryColor)">...</div>
```

In a CSS file or `<style>` block we set default values:

```css
:root {
  --primaryColor: teal;
  --secondaryColor: indianred;  
}
```

We can then use JavaScript to apply new values:

```html
<body style="--primaryColor:aliceblue;--secondaryColor:bisque;">
```

Et voilà!

In short, the idea is to **not set hard-coded color values** through ACSS, but rather relying on (CSS) variables alone.

**Reply by redonkulus** [reply]
@thierryk How do you think this applies to Dark mode? I was thinking it might be interesting if atomizer could support dark mode like Tailwind does. But not sure if atomizer should be the place for it.

**Reply by thierryk** [reply]
I don't know how Tailwind does it but it's pretty straightforward using Atomizer:

   1. You map "color keyword" to custom properties in the config, as shown above
   2. You set values in 2 rulesets, see below
   3. You toggle a class on `body` to switch between themes

```
/**
   default colors (light theme) 
*/
body {
   --primaryColor: white;
   --secondaryColor: black;
   ...
}
/**
   dark theme 
*/
body.dark {
   --primaryColor: darkgrey;
   --secondaryColor: lightgrey;
   ...
}
 ```

Then in the markup you'd go with the variables from the config:

```
<div class="C(secondaryColor) Bg(primaryColor)">...</div>
``` 

Obviously you'd use better names than `primaryColor` and `secondaryColor` for that. I just used the examples I already had in the config above. 

I guess we could come up with an easier way to do this through Atomizer where everything is set through the config (variables, custom properties, values, and the flag to switch between themes). It is Atomizer that would then write the 2 rulesets needed for the themes. 
Not sure if that would be the best approach though...

**Reply by redonkulus** [reply]
Here is the [documentation from tailwind](https://tailwindcss.com/docs/dark-mode). It's really simple to use and even customizable to work via OS setting or manual via DOM class. 

You can add a modifier to your class block for when dark mode is enabled: `text-color-black dark:text-color-white`. Then Tailwind can generate the CSS for you. 

For atomic maybe it would be some new syntax: `C(#000) Dark:C(#fff)`

**Reply by thierryk** [reply]
What's the difference (In the case of flag switching) between:

`text-color-black dark:text-color-white`

and: 

`C(#000) dark_C(#fff)` 

no need for new syntax :)

In any case, and for what it is worth, I think relying on custom properties inside a `prefers-color-scheme` MQ is better for theme switching. That way you do not need to be explicit each time you set a color via a class to set its counter part for dark theme, or else. Custom properties were designed to save you from this...

@media (prefers-color-scheme: dark) {
   /* all rules for dark theme go here */ 
...
}

As a side note, I Like how people used to say Atomizer was a bad idea because people had to remember a lot of classes and they also had to type a lot more inside `@class`. Do you see the difference between Tailwind syntax and Atomizer's syntax above? lol

**Reply by redonkulus** [reply]
@thierryk Dark mode was recently added to the Atomzier site. I also have a PR open to add a "Dark Mode" guide, let me know if anything should be added or changed https://github.com/acss-io/atomizer/pull/509

**Reply by redonkulus** [reply]
As for tailwind class names, they are more easily rememberable than atomic just cause they read like well. But after a while, you start to remember the syntax in both languages pretty quickly.

**Comment by carriemorrison** [comment]
Is there any benefit in defining custom color vars twice? we could simply use the css var in atomic.

```
<div class="C(--secondaryColor) Bg(--primaryColor)">...</div>
```

**Reply by thierryk** [reply]
No, you're right we could do it that way. 
The only benefit I see would be related to extending the syntax (via `--`). In case we need to come up with "something" to let Atomizer generate a specific styling like we do with `_` and other special characters (if that makes sense)...
